Add Aftermath mission Production Disruption

Smittytron opened the pull request:
This is one of a handful of Aftermath missions that were easy to speedrun and seemed to be designed for the sole purpose of highlighting one of the new units.

I left things as they were in the original for easy difficulty and for normal I create a few actors to prevent the chronotanks from just  jumping past the base onto the island.

Smittytron committed:
Add Aftermath mission: Production Disruption

Smittytron commented:
Rebased and Updated for #18271

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+	Trigger.OnDamaged(SubPen, function()
+		Utils.Do(Heavys, function(actor)
+			if not actor.IsDead then
+				actor.Hunt()
```
Is this supposed to only `Hunt` or should they idle-hunt?

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+
+PowerDown = false
+PowerDownTeslas = function()
+	if not PowerDown == true then
```
`if not PowerDown then`

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+		local demoEngis = Reinforcements.ReinforceWithTransport(greece, LSTType, DemoEngiTeam, DemoEngiPath, { DemoEngiPath[1] })[2][1]
+		Trigger.OnAddedToWorld(demoEngis, function()
+			CaptureRadarDome()
+		end)
```
```suggestion
		local demoEngis = Reinforcements.ReinforceWithTransport(greece, LSTType, DemoEngiTeam, DemoEngiPath, { DemoEngiPath[1] })[2]
		Trigger.OnAllRemovedFromWorld(Utils.Where(demoEngis, function(a) return a.Type == "e6"), function()
			if not DomeCaptured then
				greece.MarkFailedObjective(CaptureDome)
			end
		end)

		Trigger.OnCapture(ObjectiveDome, function()
			DomeCaptured = true
			greece.MarkCompletedObjective(CaptureDome)
			SendChronos()
		end)
```
and removing `CaptureRadarDome` should do the trick without timing hacks.

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+	Chalk1.TargetParatroopers(LZ.CenterPosition, Angle.New(740))
+	if Map.LobbyOption("difficulty") == "normal" then
+		Actor.Create("tsla", true, { Owner = ussr, Location = EasyCamera.Location })
+		Actor.Create("4tnk", true, { Owner = ussr, Facing = 127, Location = Mammoth.Location })
```
This needs to use an `Angle now`.

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+	end
+
+	if ticked > 0 then
+		UserInterface.SetMissionText("Submarine competes in " .. Utils.FormatTime(ticked), TimerColor)
```
`completes`?

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+end
+
+MissileSubEscape = function()
+	missileSub = Actor.Create("msub", true, { Owner = ussr, Location = MissileSubSpawn.Location })
```
This can be `local`.

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+
+MissileSubEscape = function()
+	missileSub = Actor.Create("msub", true, { Owner = ussr, Location = MissileSubSpawn.Location })
+	subEscapeCam = Actor.Create("camera", true, { Owner = greece, Location = SubPath2.Location })
```
This doesn't appear to be used?

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+		PowerDownTeslas()
+	end
+
+	if greece.HasNoRequiredUnits() and UnitsArrived == true then
```
`and UnitsArrived then`

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+	if ticked > 0 then
+		UserInterface.SetMissionText("Submarine competes in " .. Utils.FormatTime(ticked), TimerColor)
+		ticked = ticked - 1
+	elseif ticked == 0 and not TimerFinished == true then
```
`not TimerFinished then`

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+
+	ussrObj = ussr.AddPrimaryObjective("Get a missile sub to open waters.")
+	StopProduction = greece.AddPrimaryObjective("Destroy the Soviet sub pen.")
+	PowerDownTeslaCoils = greece.AddSecondaryObjective("Take down power to the tesla coils.")
```
This should likely be a primary objective, considering that you can't progress without doing it.

abcdefg30 commented on mods/ra/maps/production-disruption/production-disruption.lua:
```diff
@@ -0,0 +1,270 @@
+
+		Trigger.OnAllRemovedFromWorld(Engineers, function()
+			if not DomeCaptured == true then
+				greece.MarkFailedObjective(CaptureDome)
```
We need to fail the mission here in case the player hasn't repaired the bridge, as there is no way to progress.

Smittytron commented:
Updated

abcdefg30 reviewed (dismissed):
It is really hard to destroy the sub once it is built, but ok for now.

Smittytron commented:
Updated. Added a camera on the radar dome after the two flametowers are destroyed. This is due to feedback from @Orb370, who pointed out that because bridges are repairable in OpenRA, the natural assumption of a player is to repair the bridge and advance north.

abcdefg30 commented:
Merging with @Orb370's second +1.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7ea10fef4cc4b970b62472fe96d4e2dbc2fac383...d7b640ff9ea19918a97c3afa6ed16950fb03e835)
